source/adapters/level_zero/v2/memory.hpp

```diff
@@ -67,61 +75,99 @@ struct ur_integrated_mem_handle_t : public ur_mem_handle_t_ {
+  getDevicePtr(ur_device_handle_t, device_access_mode_t, size_t offset,
+               size_t size,
                std::function<void(void *src, void *dst, size_t)>) override;
-  void *mapHostPtr(access_mode_t, size_t offset, size_t size,
+  void *mapHostPtr(ur_map_flags_t, size_t offset, size_t size,
                    std::function<void(void *src, void *dst, size_t)>) override;
   void unmapHostPtr(void *pMappedPtr,
                     std::function<void(void *src, void *dst, size_t)>) override;
 
 private:
-  void *ptr;
+  std::unique_ptr<void, std::function<void(void *)>> ptr;
 };
 
 struct host_allocation_desc_t {
   host_allocation_desc_t(void *ptr, size_t size, size_t offset,
-                         ur_mem_handle_t_::access_mode_t access)
-      : ptr(ptr), size(size), offset(offset), access(access) {}
+                         ur_map_flags_t flags)
+      : ptr(ptr), size(size), offset(offset), flags(flags) {}
 
   void *ptr;
   size_t size;
   size_t offset;
-  ur_mem_handle_t_::access_mode_t access;
+  ur_map_flags_t flags;
 };
 
 // Manages memory buffer for discrete GPU.
 // Memory is allocated on the device and migrated/copies if necessary.
 struct ur_discrete_mem_handle_t : public ur_mem_handle_t_ {
+  // If hostPtr is not null, the buffer is allocated immediately on the
+  // first device in the context. Otherwise, the buffer is allocated on
+  // firt getDevicePtr call.
   ur_discrete_mem_handle_t(ur_context_handle_t hContext, void *hostPtr,
-                           size_t size);
-  ~ur_discrete_mem_handle_t();
+                           size_t size, device_access_mode_t accesMode);
+
+  // Create buffer on top of existing device memory.
+  ur_discrete_mem_handle_t(ur_context_handle_t hContext,
+                           ur_device_handle_t hDevice, void *devicePtr,
+                           size_t size, device_access_mode_t accesMode,
+                           bool ownDevicePtr);
 
   void *
-  getDevicePtr(ur_device_handle_t, access_mode_t, size_t offset, size_t size,
+  getDevicePtr(ur_device_handle_t, device_access_mode_t, size_t offset,
+               size_t size,
                std::function<void(void *src, void *dst, size_t)>) override;
-  void *mapHostPtr(access_mode_t, size_t offset, size_t size,
+  void *mapHostPtr(ur_map_flags_t, size_t offset, size_t size,
                    std::function<void(void *src, void *dst, size_t)>) override;
   void unmapHostPtr(void *pMappedPtr,
                     std::function<void(void *src, void *dst, size_t)>) override;
 
 private:
-  void *getDevicePtrUnlocked(ur_device_handle_t, access_mode_t, size_t offset,
-                             size_t size,
-                             std::function<void(void *src, void *dst, size_t)>);
+  void *getCurrentAllocation();
 
   // Vector of per-device allocations indexed by device->Id
-  std::vector<void *> deviceAllocations;
+  std::vector<std::unique_ptr<void, std::function<void(void *)>>>
```

Comment:
typedef/using? you'd be able to also shorten the inserts.

Reply:
Yeah, done.